diff --git a/ags/X/machine/ags_audiorec.c b/ags/X/machine/ags_audiorec.c
index c4b6ca888..7c94c341c 100644
--- a/ags/X/machine/ags_audiorec.c
+++ b/ags/X/machine/ags_audiorec.c
@@ -29,6 +29,8 @@
 
 #include <ags/X/thread/ags_gui_thread.h>
 
+#include <ags/i18n.h>
+
 void ags_audiorec_class_init(AgsAudiorecClass *audiorec);
 void ags_audiorec_connectable_interface_init(AgsConnectableInterface *connectable);
 void ags_audiorec_plugin_interface_init(AgsPluginInterface *plugin);
@@ -46,6 +48,7 @@ void ags_audiorec_read(AgsFile *file, xmlNode *node, AgsPlugin *plugin);
 xmlNode* ags_audiorec_write(AgsFile *file, xmlNode *parent, AgsPlugin *plugin);
 
 void ags_audiorec_show(GtkWidget *widget);
+void ags_audiorec_show_all(GtkWidget *widget);
 
 void ags_audiorec_map_recall(AgsMachine *machine);
 
@@ -72,6 +75,7 @@ void ags_audiorec_input_map_recall(AgsAudiorec *audiorec, guint input_pad_start)
 static gpointer ags_audiorec_parent_class = NULL;
 static AgsConnectableInterface *ags_audiorec_parent_connectable_interface;
 
+GHashTable *ags_audiorec_wave_loader_completed = NULL;
 GHashTable *ags_audiorec_indicator_queue_draw = NULL;
 
 GType
@@ -142,6 +146,7 @@ ags_audiorec_class_init(AgsAudiorecClass *audiorec)
   widget = (GtkWidgetClass *) audiorec;
 
   widget->show = ags_audiorec_show;
+  widget->show_all = ags_audiorec_show_all;
 
   /* AgsMachineClass */
   machine = (AgsMachineClass *) audiorec;
@@ -221,7 +226,7 @@ ags_audiorec_init(AgsAudiorec *audiorec)
 		    (GtkWidget *) hbox);
 
   /* frame - filename and open */
-  frame = (GtkFrame *) gtk_frame_new("file");
+  frame = (GtkFrame *) gtk_frame_new(i18n("file"));
   gtk_box_pack_start(hbox,
 		     frame,
 		     FALSE, FALSE,
@@ -240,7 +245,7 @@ ags_audiorec_init(AgsAudiorec *audiorec)
 		     FALSE, FALSE,
 		     0);
   
-  label = gtk_label_new("filename: ");
+  label = gtk_label_new(i18n("filename: "));
   gtk_box_pack_start(filename_hbox,
 		     label,
 		     FALSE, FALSE,
@@ -257,6 +262,19 @@ ags_audiorec_init(AgsAudiorec *audiorec)
 		     audiorec->open,
 		     FALSE, FALSE,
 		     0);
+
+  audiorec->wave_loader = NULL;
+
+  audiorec->position = -1;
+
+  audiorec->loading = gtk_label_new(i18n("loading ...  "));
+  gtk_box_pack_start(filename_hbox,
+		     audiorec->loading,
+		     FALSE, FALSE,
+		     0);
+  gtk_widget_set_no_show_all(audiorec->loading,
+			     TRUE);
+  gtk_widget_hide(audiorec->loading);
   
   /* radio */
   radio_hbox = (GtkHBox *) gtk_hbox_new(FALSE,
@@ -288,7 +306,7 @@ ags_audiorec_init(AgsAudiorec *audiorec)
 		     0);
   
   /* frame - hindicator */
-  frame = (GtkFrame *) gtk_frame_new("input");
+  frame = (GtkFrame *) gtk_frame_new(i18n("input"));
   gtk_box_pack_start(hbox,
 		     frame,
 		     FALSE, FALSE,
@@ -302,6 +320,17 @@ ags_audiorec_init(AgsAudiorec *audiorec)
   /* dialog */
   audiorec->open_dialog = NULL;
 
+  /* wave_loader */
+  if(ags_audiorec_wave_loader_completed == NULL){
+    ags_audiorec_wave_loader_completed = g_hash_table_new_full(g_direct_hash, g_direct_equal,
+							       NULL,
+							       NULL);
+  }
+
+  g_hash_table_insert(ags_audiorec_wave_loader_completed,
+		      audiorec, ags_audiorec_wave_loader_completed_timeout);
+  g_timeout_add(1000 / 4, (GSourceFunc) ags_audiorec_wave_loader_completed_timeout, (gpointer) audiorec);
+
   /* indicator */
   if(ags_audiorec_indicator_queue_draw == NULL){
     ags_audiorec_indicator_queue_draw = g_hash_table_new_full(g_direct_hash, g_direct_equal,
@@ -317,6 +346,16 @@ ags_audiorec_init(AgsAudiorec *audiorec)
 void
 ags_audiorec_finalize(GObject *gobject)
 {
+  AgsAudiorec *audiorec;
+
+  audiorec = gobject;
+
+  g_hash_table_remove(ags_audiorec_wave_loader_completed,
+		      audiorec);
+  
+  g_hash_table_remove(ags_audiorec_indicator_queue_draw,
+		      audiorec);
+  
   /* call parent */
   G_OBJECT_CLASS(ags_audiorec_parent_class)->finalize(gobject);
 }
@@ -422,9 +461,25 @@ ags_audiorec_set_xml_type(AgsPlugin *plugin, gchar *xml_type)
 void
 ags_audiorec_show(GtkWidget *widget)
 {
+  AgsAudiorec *audiorec;
+
+  audiorec = widget;
+  
+  /* call parent */
   GTK_WIDGET_CLASS(ags_audiorec_parent_class)->show(widget);
 }
 
+void
+ags_audiorec_show_all(GtkWidget *widget)
+{
+  AgsAudiorec *audiorec;
+
+  audiorec = widget;
+  
+  /* call parent */
+  GTK_WIDGET_CLASS(ags_audiorec_parent_class)->show_all(widget);
+}
+
 void
 ags_audiorec_map_recall(AgsMachine *machine)
 {
@@ -597,6 +652,8 @@ ags_audiorec_resize_audio_channels(AgsMachine *machine,
 				    audio_channels_old);
 
       while(channel != next_pad){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_WAVE));
+	
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -727,6 +784,8 @@ ags_audiorec_resize_pads(AgsMachine *machine, GType type,
 				    pads_old);
 
       while(channel != NULL){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_WAVE));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -848,61 +907,85 @@ void
 ags_audiorec_open_filename(AgsAudiorec *audiorec,
 			   gchar *filename)
 {
-  AgsAudioFile *audio_file;
-
-  AgsOpenWave *open_wave;
-
-  AgsGuiThread *gui_thread;
-  
-  AgsApplicationContext *application_context;
-  
-  GObject *output_soundcard;
-
-  GList *start_wave, *wave;
+  AgsWaveLoader *wave_loader;
   
   if(!AGS_IS_AUDIOREC(audiorec) ||
      filename == NULL){
     return;
   }
 
-  application_context = ags_application_context_get_instance();
-
-  gui_thread = ags_ui_provider_get_gui_thread(AGS_UI_PROVIDER(application_context));
-  
-  g_object_get(AGS_MACHINE(audiorec)->audio,
-	       "output-soundcard", &output_soundcard,
-	       NULL);
-
-  g_object_get(AGS_MACHINE(audiorec)->audio,
-	       "wave", &start_wave,
-	       NULL);
+  audiorec->wave_loader = 
+    wave_loader = ags_wave_loader_new(AGS_MACHINE(audiorec)->audio,
+				      filename,
+				      TRUE);
 
-  wave = start_wave;
+  ags_wave_loader_start(wave_loader);
+}
 
-  while(wave != NULL){
-    ags_audio_remove_wave(AGS_MACHINE(audiorec)->audio,
-			  wave->data);
+/**
+ * ags_audiorec_wave_loader_completed_timeout:
+ * @widget: the widget
+ *
+ * Queue draw widget
+ *
+ * Returns: %TRUE if proceed poll completed, otherwise %FALSE
+ *
+ * Since: 2.0.13
+ */
+gboolean
+ags_audiorec_wave_loader_completed_timeout(AgsAudiorec *audiorec)
+{
+  if(g_hash_table_lookup(ags_audiorec_wave_loader_completed,
+			 audiorec) != NULL){
+    if(audiorec->wave_loader != NULL){
+      if(ags_wave_loader_test_flags(audiorec->wave_loader, AGS_WAVE_LOADER_HAS_COMPLETED)){
+	g_object_run_dispose(audiorec->wave_loader);
+	g_object_unref(audiorec->wave_loader);
+
+	audiorec->wave_loader = NULL;
+
+	audiorec->position = -1;
+	gtk_widget_hide(audiorec->loading);
+      }else{
+	if(audiorec->position == -1){
+	  audiorec->position = 0;
+
+	  gtk_widget_show(audiorec->loading);
+	}
+
+	switch(audiorec->position){
+	case 0:
+	  {
+	    audiorec->position = 1;
+	    
+	    gtk_label_set_label(audiorec->loading,
+				"loading ...  ");
+	  }
+	  break;
+	case 1:
+	  {
+	    audiorec->position = 2;
+
+	    gtk_label_set_label(audiorec->loading,
+				"loading  ... ");
+	  }
+	  break;
+	case 2:
+	  {
+	    audiorec->position = 0;
+
+	    gtk_label_set_label(audiorec->loading,
+				"loading   ...");
+	  }
+	  break;
+	}
+      }
+    }
     
-    wave = wave->next;
+    return(TRUE);
+  }else{
+    return(FALSE);
   }
-
-  g_list_free(start_wave);
-  
-  audio_file = ags_audio_file_new(filename,
-				  output_soundcard,
-				  -1);
-  ags_audio_file_open(audio_file);
-
-  g_object_set(AGS_MACHINE(audiorec)->audio,
-	       "input-audio-file", audio_file,
-	       NULL);
-
-  open_wave = ags_open_wave_new(AGS_MACHINE(audiorec)->audio,
-				audio_file,
-				NULL,
-				0);
-  ags_gui_thread_schedule_task(gui_thread,
-			       open_wave);
 }
 
 /**
diff --git a/ags/X/machine/ags_audiorec.h b/ags/X/machine/ags_audiorec.h
index b2d1ec47b..bf6b8396b 100644
--- a/ags/X/machine/ags_audiorec.h
+++ b/ags/X/machine/ags_audiorec.h
@@ -49,10 +49,15 @@ struct _AgsAudiorec
 
   gchar *name;
   gchar *xml_type;
-
+  
   GtkEntry *filename;
   GtkButton *open;
 
+  AgsWaveLoader *wave_loader;
+
+  gint position;
+  GtkLabel *loading;
+
   GtkRadioButton *keep_data;
   GtkRadioButton *mix_data;
   GtkRadioButton *replace_data;
@@ -72,6 +77,7 @@ GType ags_audiorec_get_type(void);
 void ags_audiorec_open_filename(AgsAudiorec *audiorec,
 				gchar *filename);
 
+gboolean ags_audiorec_wave_loader_completed_timeout(AgsAudiorec *audiorec);
 gboolean ags_audiorec_indicator_queue_draw_timeout(AgsAudiorec *audiorec);
 
 AgsAudiorec* ags_audiorec_new(GObject *soundcard);
diff --git a/ags/X/machine/ags_drum.c b/ags/X/machine/ags_drum.c
index 3f1db1604..56fb0d91c 100644
--- a/ags/X/machine/ags_drum.c
+++ b/ags/X/machine/ags_drum.c
@@ -207,7 +207,8 @@ ags_drum_init(AgsDrum *drum)
 			      AGS_AUDIO_OUTPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_FILE));
-  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_SEQUENCER |
+  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_PLAYBACK |
+				      AGS_SOUND_ABILITY_SEQUENCER |
 				      AGS_SOUND_ABILITY_NOTATION));
   ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_PATTERN_MODE |
 					AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
diff --git a/ags/X/machine/ags_dssi_bridge.c b/ags/X/machine/ags_dssi_bridge.c
index 98f9d0930..da8e9f30f 100644
--- a/ags/X/machine/ags_dssi_bridge.c
+++ b/ags/X/machine/ags_dssi_bridge.c
@@ -267,7 +267,8 @@ ags_dssi_bridge_init(AgsDssiBridge *dssi_bridge)
 			      AGS_AUDIO_ASYNC |
 			      AGS_AUDIO_OUTPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_RECYCLING));
-  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_NOTATION));
+  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_PLAYBACK |
+				      AGS_SOUND_ABILITY_NOTATION));
   ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
 					AGS_SOUND_BEHAVIOUR_DEFAULTS_TO_INPUT));
   g_object_set(audio,
@@ -811,6 +812,8 @@ ags_dssi_bridge_resize_audio_channels(AgsMachine *machine,
 				audio_channels_old);
       
       while(channel != next_pad){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+	
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -971,6 +974,8 @@ ags_dssi_bridge_resize_pads(AgsMachine *machine, GType type,
 				    pads_old);
 
       while(channel != NULL){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -1038,6 +1043,7 @@ ags_dssi_bridge_map_recall(AgsMachine *machine)
   dssi_bridge = (AgsDssiBridge *) machine;
 
   audio = machine->audio;
+
   /* ags-delay */
   ags_recall_factory_create(audio,
 			    NULL, NULL,
@@ -1098,111 +1104,111 @@ ags_dssi_bridge_map_recall(AgsMachine *machine)
   }
 
   g_list_free(start_play);
-  
-  /* ags-route-dssi */
+    
+  /* ags-record-midi */
   ags_recall_factory_create(audio,
 			    NULL, NULL,
-			    "ags-route-dssi",
+			    "ags-record-midi",
 			    0, 0,
 			    0, 0,
 			    (AGS_RECALL_FACTORY_INPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
-	       "recall", &start_recall,
+	       "play", &start_play,
 	       NULL);
   
-  recall = ags_recall_find_type(start_recall,
-				AGS_TYPE_ROUTE_DSSI_AUDIO_RUN);
-
-  if(recall != NULL){
-    recall_route_dssi_audio_run = AGS_ROUTE_DSSI_AUDIO_RUN(recall->data);
+  play = ags_recall_find_type(start_play,
+			      AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
 
+  if(play != NULL){
+    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(play->data);
+    
     /* set dependency */
-    g_object_set(G_OBJECT(recall_route_dssi_audio_run),
+    g_object_set(G_OBJECT(recall_record_midi_audio_run),
 		 "delay-audio-run", play_delay_audio_run,
 		 NULL);
 
     /* set dependency */
-    g_object_set(G_OBJECT(recall_route_dssi_audio_run),
+    g_object_set(G_OBJECT(recall_record_midi_audio_run),
 		 "count-beats-audio-run", play_count_beats_audio_run,
 		 NULL);
-  }else{
-    recall_route_dssi_audio_run = NULL;
-  }
+  }  
 
-  g_list_free(start_recall);
+  g_list_free(start_play);
   
-  /* ags-record-midi */
+  /* ags-play-notation */
   ags_recall_factory_create(audio,
 			    NULL, NULL,
-			    "ags-record-midi",
+			    "ags-play-notation",
 			    0, 0,
 			    0, 0,
 			    (AGS_RECALL_FACTORY_INPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
-	       "recall", &start_recall,
+	       "play", &start_play,
 	       NULL);
   
-  recall = ags_recall_find_type(start_recall,
-				AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
+  play = ags_recall_find_type(start_play,
+			      AGS_TYPE_PLAY_NOTATION_AUDIO_RUN);
+
+  if(play != NULL){
+    recall_notation_audio_run = AGS_PLAY_NOTATION_AUDIO_RUN(play->data);
 
-  if(recall != NULL){
-    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(recall->data);
-    
     /* set dependency */
-    g_object_set(G_OBJECT(recall_record_midi_audio_run),
+    g_object_set(G_OBJECT(recall_notation_audio_run),
 		 "delay-audio-run", play_delay_audio_run,
 		 NULL);
 
     /* set dependency */
-    g_object_set(G_OBJECT(recall_record_midi_audio_run),
+    g_object_set(G_OBJECT(recall_notation_audio_run),
 		 "count-beats-audio-run", play_count_beats_audio_run,
 		 NULL);
-  }  
+  }
 
-  g_list_free(start_recall);
+  g_list_free(start_play);
   
-  /* ags-play-notation */
+  /* ags-route-dssi */
   ags_recall_factory_create(audio,
 			    NULL, NULL,
-			    "ags-play-notation",
+			    "ags-route-dssi",
 			    0, 0,
 			    0, 0,
 			    (AGS_RECALL_FACTORY_INPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
-	       "recall", &start_recall,
+	       "play", &start_play,
 	       NULL);
   
-  recall = ags_recall_find_type(start_recall,
-				AGS_TYPE_PLAY_NOTATION_AUDIO_RUN);
+  play = ags_recall_find_type(start_play,
+			      AGS_TYPE_ROUTE_DSSI_AUDIO_RUN);
 
-  if(recall != NULL){
-    recall_notation_audio_run = AGS_PLAY_NOTATION_AUDIO_RUN(recall->data);
+  if(play != NULL){
+    recall_route_dssi_audio_run = AGS_ROUTE_DSSI_AUDIO_RUN(play->data);
 
     /* set dependency */
-    g_object_set(G_OBJECT(recall_notation_audio_run),
+    g_object_set(G_OBJECT(recall_route_dssi_audio_run),
 		 "delay-audio-run", play_delay_audio_run,
 		 NULL);
 
     /* set dependency */
-    g_object_set(G_OBJECT(recall_notation_audio_run),
+    g_object_set(G_OBJECT(recall_route_dssi_audio_run),
 		 "count-beats-audio-run", play_count_beats_audio_run,
 		 NULL);
+  }else{
+    recall_route_dssi_audio_run = NULL;
   }
 
-  g_list_free(start_recall);
-  
+  g_list_free(start_play);
+
   /* depending on destination */
   ags_dssi_bridge_input_map_recall(dssi_bridge,
 				   0,
diff --git a/ags/X/machine/ags_ffplayer.c b/ags/X/machine/ags_ffplayer.c
index b94e1d5bb..6b4be4dec 100644
--- a/ags/X/machine/ags_ffplayer.c
+++ b/ags/X/machine/ags_ffplayer.c
@@ -201,7 +201,8 @@ ags_ffplayer_init(AgsFFPlayer *ffplayer)
 			      AGS_AUDIO_OUTPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_FILE));
-  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_NOTATION));
+  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_PLAYBACK |
+				      AGS_SOUND_ABILITY_NOTATION));
   ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
 					AGS_SOUND_BEHAVIOUR_DEFAULTS_TO_INPUT));
   g_object_set(audio,
@@ -791,19 +792,67 @@ ags_ffplayer_resize_audio_channels(AgsMachine *machine,
 				   guint audio_channels, guint audio_channels_old,
 				   gpointer data)
 {
-  AgsAudio *audio;
+  AgsAudio *audio;  
+  AgsChannel *output;
+  AgsChannel *channel, *next_pad;
   
-  guint input_pads;
+  guint output_pads, input_pads;
 
   audio = machine->audio;
 
   /* get some fields */
   g_object_get(audio,
 	       "input-pads", &input_pads,
+	       "output-pads", &output_pads,
+	       "output", &output,
 	       NULL);
 
   /*  */
   if(audio_channels > audio_channels_old){  
+    /* AgsOutput */
+    channel = output;
+
+    while(channel != NULL){
+      /* get some fields */
+      g_object_get(channel,
+		   "next-pad", &next_pad,
+		   NULL);
+
+      channel = ags_channel_pad_nth(channel,
+				    audio_channels_old);
+
+      while(channel != next_pad){
+	AgsRecycling *recycling;
+	AgsAudioSignal *audio_signal;
+
+	GObject *output_soundcard;
+
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+		
+	g_object_get(audio,
+		     "output-soundcard", &output_soundcard,
+		     NULL);
+
+	/* get recycling */
+	g_object_get(channel,
+		     "first-recycling", &recycling,
+		     NULL);
+
+	/* instantiate template audio signal */
+	audio_signal = ags_audio_signal_new(output_soundcard,
+					    (GObject *) recycling,
+					    NULL);
+	audio_signal->flags |= AGS_AUDIO_SIGNAL_TEMPLATE;
+	ags_recycling_add_audio_signal(recycling,
+				       audio_signal);
+
+	/* iterate */
+	g_object_get(channel,
+		     "next", &channel,
+		     NULL);
+      }
+    }
+
     if(ags_recall_global_get_rt_safe() ||
        ags_recall_global_get_performance_mode()){
       /* ags-copy */
@@ -903,6 +952,8 @@ ags_ffplayer_resize_pads(AgsMachine *machine, GType channel_type,
   AgsFFPlayer *ffplayer;
 
   AgsAudio *audio;
+  AgsChannel *output;
+  AgsChannel *channel;
 
   guint output_pads, input_pads;
   gboolean grow;
@@ -919,6 +970,7 @@ ags_ffplayer_resize_pads(AgsMachine *machine, GType channel_type,
   g_object_get(audio,
 	       "output-pads", &output_pads,
 	       "input-pads", &input_pads,
+	       "output", &output,
 	       NULL);
 
   /* check grow */
@@ -937,6 +989,19 @@ ags_ffplayer_resize_pads(AgsMachine *machine, GType channel_type,
     }
   }else if(channel_type == AGS_TYPE_OUTPUT){
     if(grow){
+      /* AgsOutput */
+      channel = ags_channel_pad_nth(output,
+				    pads_old);
+
+      while(channel != NULL){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
+	/* iterate */
+	g_object_get(channel,
+		     "next", &channel,
+		     NULL);
+      }
+
       /* depending on destination */
       ags_ffplayer_output_map_recall(ffplayer, pads_old);
     }else{
diff --git a/ags/X/machine/ags_live_dssi_bridge.c b/ags/X/machine/ags_live_dssi_bridge.c
index 0ff805e9f..0d1d8be38 100644
--- a/ags/X/machine/ags_live_dssi_bridge.c
+++ b/ags/X/machine/ags_live_dssi_bridge.c
@@ -273,7 +273,8 @@ ags_live_dssi_bridge_init(AgsLiveDssiBridge *live_dssi_bridge)
 			      AGS_AUDIO_INPUT_HAS_RECYCLING |
 			      AGS_AUDIO_SKIP_INPUT));
   ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_NOTATION));
-  ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING));
+  ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
+					AGS_SOUND_BEHAVIOUR_DEFAULTS_TO_INPUT));
   //  audio->flags &= (~AGS_AUDIO_NOTATION_DEFAULT);
   
   g_object_set(audio,
@@ -812,6 +813,8 @@ ags_live_dssi_bridge_resize_audio_channels(AgsMachine *machine,
 				audio_channels_old);
 
       while(channel != next_pad){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -969,6 +972,8 @@ ags_live_dssi_bridge_resize_pads(AgsMachine *machine, GType type,
 				    pads_old);
 
       while(channel != NULL){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -1043,8 +1048,7 @@ ags_live_dssi_bridge_map_recall(AgsMachine *machine)
 			    0, 0,
 			    (AGS_RECALL_FACTORY_OUTPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_PLAY |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
@@ -1072,7 +1076,7 @@ ags_live_dssi_bridge_map_recall(AgsMachine *machine)
 			    (AGS_RECALL_FACTORY_OUTPUT |
 			     AGS_RECALL_FACTORY_ADD |
 			     AGS_RECALL_FACTORY_PLAY |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
@@ -1106,18 +1110,18 @@ ags_live_dssi_bridge_map_recall(AgsMachine *machine)
 			    0, 0,
 			    (AGS_RECALL_FACTORY_INPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
-	       "recall", &start_recall,
+	       "play", &start_play,
 	       NULL);
 
-  recall = ags_recall_find_type(start_recall,
-				AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
+  play = ags_recall_find_type(start_play,
+			      AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
 
-  if(recall != NULL){
-    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(recall->data);
+  if(play != NULL){
+    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(play->data);
     
     /* set dependency */
     g_object_set(G_OBJECT(recall_record_midi_audio_run),
@@ -1130,7 +1134,7 @@ ags_live_dssi_bridge_map_recall(AgsMachine *machine)
 		 NULL);
   }  
 
-  g_list_free(start_recall);
+  g_list_free(start_play);
   
   /* ags-play-dssi */
   ags_recall_factory_create(audio,
@@ -1231,6 +1235,8 @@ ags_live_dssi_bridge_output_map_recall(AgsLiveDssiBridge *live_dssi_bridge,
 void
 ags_live_dssi_bridge_load(AgsLiveDssiBridge *live_dssi_bridge)
 {
+  AgsEffectBulk *effect_bulk;
+  AgsBulkMember *bulk_member;
   GtkListStore *model;
 
   GtkTreeIter iter;
@@ -1244,8 +1250,9 @@ ags_live_dssi_bridge_load(AgsLiveDssiBridge *live_dssi_bridge)
   LADSPA_PortDescriptor *port_descriptor;
   LADSPA_PortRangeHintDescriptor hint_descriptor;
 
+  GList *start_list, *list;
   GList *start_plugin_port, *plugin_port;
-
+  
   unsigned long samplerate;
   unsigned long effect_index;
   gdouble step;
@@ -1348,6 +1355,226 @@ ags_live_dssi_bridge_load(AgsLiveDssiBridge *live_dssi_bridge)
   
   gtk_combo_box_set_model(GTK_COMBO_BOX(live_dssi_bridge->program),
 			  GTK_TREE_MODEL(model));
+
+  effect_bulk = AGS_EFFECT_BRIDGE(AGS_MACHINE(live_dssi_bridge)->bridge)->bulk_output;
+
+  /* retrieve position within table  */
+  x = 0;
+  y = 0;
+  
+  list = effect_bulk->table->children;
+
+  while(list != NULL){
+    if(y <= ((GtkTableChild *) list->data)->top_attach){
+      y = ((GtkTableChild *) list->data)->top_attach + 1;
+    }
+
+    list = list->next;
+  }
+
+  /* load ports */
+  g_object_get(dssi_plugin,
+	       "plugin-port", &start_plugin_port,
+	       NULL);
+  
+  plugin_port = start_plugin_port;
+
+  port_count = g_list_length(start_plugin_port);
+  k = 0;
+
+  while(plugin_port != NULL){
+    if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_CONTROL)){
+      GtkWidget *child_widget;
+
+      AgsLadspaConversion *ladspa_conversion;
+      
+      GType widget_type;
+
+      gchar *plugin_name;
+      gchar *control_port;
+      
+      guint step_count;
+      gboolean disable_seemless;
+
+      disable_seemless = FALSE;
+      
+      if(x == AGS_EFFECT_BULK_COLUMNS_COUNT){
+	x = 0;
+	y++;
+	gtk_table_resize(effect_bulk->table,
+			 y + 1, AGS_EFFECT_BULK_COLUMNS_COUNT);
+      }
+
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_TOGGLED)){
+	if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_OUTPUT)){
+	  widget_type = AGS_TYPE_LED;
+	}else{
+	  widget_type = GTK_TYPE_TOGGLE_BUTTON;
+	}
+      }else{
+	if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_OUTPUT)){
+	  widget_type = AGS_TYPE_HINDICATOR;
+	}else{
+	  widget_type = AGS_TYPE_DIAL;
+	}
+      }
+
+      step_count = AGS_DIAL_DEFAULT_PRECISION;
+
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_INTEGER)){
+	step_count = AGS_PLUGIN_PORT(plugin_port->data)->scale_steps;
+
+	disable_seemless = TRUE;	
+      }
+
+      /* add bulk member */
+      plugin_name = g_strdup_printf("dssi-%u",
+				    dssi_plugin->unique_id);
+      control_port = g_strdup_printf("%u/%u",
+				     k,
+				     port_count);
+      bulk_member = (AgsBulkMember *) g_object_new(AGS_TYPE_BULK_MEMBER,
+						   "widget-type", widget_type,
+						   "widget-label", AGS_PLUGIN_PORT(plugin_port->data)->port_name,
+						   "plugin-name", plugin_name,
+						   "filename", live_dssi_bridge->filename,
+						   "effect", live_dssi_bridge->effect,
+						   "specifier", AGS_PLUGIN_PORT(plugin_port->data)->port_name,
+						   "control-port", control_port,
+						   "steps", step_count,
+						   NULL);
+      child_widget = ags_bulk_member_get_widget(bulk_member);
+
+      g_free(plugin_name);
+      g_free(control_port);
+      
+      /* ladspa conversion */
+      ladspa_conversion = NULL;
+
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_BOUNDED_BELOW)){
+	if(ladspa_conversion == NULL ||
+	   !AGS_IS_LADSPA_CONVERSION(ladspa_conversion)){
+	  ladspa_conversion = ags_ladspa_conversion_new();
+	}
+
+	ladspa_conversion->flags |= AGS_LADSPA_CONVERSION_BOUNDED_BELOW;
+      }
+
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_BOUNDED_ABOVE)){
+	if(ladspa_conversion == NULL ||
+	   !AGS_IS_LADSPA_CONVERSION(ladspa_conversion)){
+	  ladspa_conversion = ags_ladspa_conversion_new();
+	}
+
+	ladspa_conversion->flags |= AGS_LADSPA_CONVERSION_BOUNDED_ABOVE;
+      }
+      
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_SAMPLERATE)){
+	if(ladspa_conversion == NULL ||
+	   !AGS_IS_LADSPA_CONVERSION(ladspa_conversion)){
+	  ladspa_conversion = ags_ladspa_conversion_new();
+	}
+
+	ladspa_conversion->flags |= AGS_LADSPA_CONVERSION_SAMPLERATE;
+      }
+
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_LOGARITHMIC)){
+	if(ladspa_conversion == NULL ||
+	   !AGS_IS_LADSPA_CONVERSION(ladspa_conversion)){
+	  ladspa_conversion = ags_ladspa_conversion_new();
+	}
+    
+	ladspa_conversion->flags |= AGS_LADSPA_CONVERSION_LOGARITHMIC;
+      }
+
+      bulk_member->conversion = (AgsConversion *) ladspa_conversion;
+
+      /* child widget */
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_TOGGLED)){
+	bulk_member->port_flags = AGS_BULK_MEMBER_PORT_BOOLEAN;
+      }
+      
+      if(ags_plugin_port_test_flags(plugin_port->data, AGS_PLUGIN_PORT_INTEGER)){
+	bulk_member->port_flags = AGS_BULK_MEMBER_PORT_INTEGER;
+      }
+
+      if(AGS_IS_DIAL(child_widget)){
+	AgsDial *dial;
+	GtkAdjustment *adjustment;
+
+	LADSPA_Data lower_bound, upper_bound;
+	LADSPA_Data default_value;
+	
+	dial = (AgsDial *) child_widget;
+
+	if(disable_seemless){
+	  dial->flags &= (~AGS_DIAL_SEEMLESS_MODE);
+	}
+
+	/* add controls of ports and apply range  */
+	lower_bound = g_value_get_float(AGS_PLUGIN_PORT(plugin_port->data)->lower_value);
+	upper_bound = g_value_get_float(AGS_PLUGIN_PORT(plugin_port->data)->upper_value);
+
+	adjustment = (GtkAdjustment *) gtk_adjustment_new(0.0, 0.0, 1.0, 0.1, 0.1, 0.0);
+	g_object_set(dial,
+		     "adjustment", adjustment,
+		     NULL);
+
+	if(upper_bound >= 0.0 && lower_bound >= 0.0){
+	  step = (upper_bound - lower_bound) / step_count;
+	}else if(upper_bound < 0.0 && lower_bound < 0.0){
+	  step = -1.0 * (lower_bound - upper_bound) / step_count;
+	}else{
+	  step = (upper_bound - lower_bound) / step_count;
+	}
+
+	gtk_adjustment_set_step_increment(adjustment,
+					  step);
+	gtk_adjustment_set_lower(adjustment,
+				 lower_bound);
+	gtk_adjustment_set_upper(adjustment,
+				 upper_bound);
+
+	default_value = (LADSPA_Data) g_value_get_float(AGS_PLUGIN_PORT(plugin_port->data)->default_value);
+
+	if(ladspa_conversion != NULL){
+	  //	  default_value = ags_ladspa_conversion_convert(ladspa_conversion,
+	  //						default_value,
+	  //						TRUE);
+	}
+	
+	gtk_adjustment_set_value(adjustment,
+				 default_value);
+
+#ifdef AGS_DEBUG
+	g_message("dssi bounds: %f %f", lower_bound, upper_bound);
+#endif
+      }else if(AGS_IS_INDICATOR(child_widget) ||
+	       AGS_IS_LED(child_widget)){
+	g_hash_table_insert(ags_effect_bulk_indicator_queue_draw,
+			    child_widget, ags_effect_bulk_indicator_queue_draw_timeout);
+	effect_bulk->queued_drawing = g_list_prepend(effect_bulk->queued_drawing,
+						     child_widget);
+	g_timeout_add(1000 / 30, (GSourceFunc) ags_effect_bulk_indicator_queue_draw_timeout, (gpointer) child_widget);
+      }
+
+      gtk_table_attach(effect_bulk->table,
+		       (GtkWidget *) bulk_member,
+		       x, x + 1,
+		       y, y + 1,
+		       GTK_FILL, GTK_FILL,
+		       0, 0);
+      ags_connectable_connect(AGS_CONNECTABLE(bulk_member));
+      gtk_widget_show_all((GtkWidget *) effect_bulk->table);
+
+      x++;
+    }
+
+    plugin_port = plugin_port->next;    
+    k++;
+  }
+
+  g_list_free(start_plugin_port);
 }
 
 /**
diff --git a/ags/X/machine/ags_live_lv2_bridge.c b/ags/X/machine/ags_live_lv2_bridge.c
index dfd14c829..cda246b64 100644
--- a/ags/X/machine/ags_live_lv2_bridge.c
+++ b/ags/X/machine/ags_live_lv2_bridge.c
@@ -364,7 +364,8 @@ ags_live_lv2_bridge_init(AgsLiveLv2Bridge *live_lv2_bridge)
 			      AGS_AUDIO_INPUT_HAS_RECYCLING |
 			      AGS_AUDIO_SKIP_INPUT));
   ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_NOTATION));
-  ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING));
+  ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
+					AGS_SOUND_BEHAVIOUR_DEFAULTS_TO_INPUT));
   //  audio->flags &= (~AGS_AUDIO_NOTATION_DEFAULT);
   
   g_object_set(audio,
@@ -1005,6 +1006,8 @@ ags_live_lv2_bridge_resize_audio_channels(AgsMachine *machine,
 				    audio_channels_old);
 
       while(channel != next_pad){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -1126,6 +1129,8 @@ ags_live_lv2_bridge_resize_pads(AgsMachine *machine, GType channel_type,
 				    pads_old);
 
       while(channel != NULL){
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	/* get some fields */
 	g_object_get(channel,
 		     "output-soundcard", &output_soundcard,
@@ -1200,8 +1205,7 @@ ags_live_lv2_bridge_map_recall(AgsMachine *machine)
 			    0, 0,
 			    (AGS_RECALL_FACTORY_OUTPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_PLAY |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
@@ -1228,8 +1232,7 @@ ags_live_lv2_bridge_map_recall(AgsMachine *machine)
 			    0, 0,
 			    (AGS_RECALL_FACTORY_OUTPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_PLAY |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
   
   g_object_get(audio,
@@ -1263,19 +1266,19 @@ ags_live_lv2_bridge_map_recall(AgsMachine *machine)
 			    0, 0,
 			    (AGS_RECALL_FACTORY_INPUT |
 			     AGS_RECALL_FACTORY_ADD |
-			     AGS_RECALL_FACTORY_RECALL),
+			     AGS_RECALL_FACTORY_PLAY),
 			    0);
 
   g_object_get(audio,
-	       "recall", &start_recall,
+	       "play", &start_play,
 	       NULL);
 
-  recall = ags_recall_find_type(start_recall,
-				AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
+  play = ags_recall_find_type(start_play,
+			      AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
 
-  if(recall != NULL){
-    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(recall->data);
-    
+  if(play != NULL){
+    recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(play->data);
+   
     /* set dependency */
     g_object_set(G_OBJECT(recall_record_midi_audio_run),
 		 "delay-audio-run", play_delay_audio_run,
@@ -1287,7 +1290,7 @@ ags_live_lv2_bridge_map_recall(AgsMachine *machine)
 		 NULL);
   }  
 
-  g_list_free(start_recall);
+  g_list_free(start_play);
   
   /* ags-play-lv2 */
   ags_recall_factory_create(audio,
@@ -1681,6 +1684,8 @@ ags_live_lv2_bridge_load(AgsLiveLv2Bridge *live_lv2_bridge)
 	       "plugin-port", &start_plugin_port,
 	       NULL);
 
+  plugin_port = start_plugin_port;
+  
   port_count = g_list_length(start_plugin_port);
   k = 0;
 
diff --git a/ags/X/machine/ags_lv2_bridge.c b/ags/X/machine/ags_lv2_bridge.c
index 95dd0ca9e..6608dab5f 100644
--- a/ags/X/machine/ags_lv2_bridge.c
+++ b/ags/X/machine/ags_lv2_bridge.c
@@ -1039,6 +1039,9 @@ ags_lv2_bridge_resize_audio_channels(AgsMachine *machine,
 		     NULL);
 
 	if(first_recycling != NULL){
+	  ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_PLAYBACK |
+						  AGS_SOUND_ABILITY_NOTATION));
+
 	  /* audio signal */
 	  audio_signal = ags_audio_signal_new(output_soundcard,
 					      (GObject *) first_recycling,
@@ -1166,6 +1169,8 @@ ags_lv2_bridge_resize_pads(AgsMachine *machine, GType channel_type,
 		     NULL);
 
 	if(first_recycling != NULL){
+	  ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+
 	  /* audio signal */
 	  audio_signal = ags_audio_signal_new(output_soundcard,
 					      (GObject *) first_recycling,
@@ -1239,8 +1244,7 @@ ags_lv2_bridge_map_recall(AgsMachine *machine)
 			      0, 0,
 			      (AGS_RECALL_FACTORY_OUTPUT |
 			       AGS_RECALL_FACTORY_ADD |
-			       AGS_RECALL_FACTORY_PLAY |
-			       AGS_RECALL_FACTORY_RECALL),
+			       AGS_RECALL_FACTORY_PLAY),
 			      0);
 
     g_object_get(audio,
@@ -1267,8 +1271,7 @@ ags_lv2_bridge_map_recall(AgsMachine *machine)
 			      0, 0,
 			      (AGS_RECALL_FACTORY_OUTPUT |
 			       AGS_RECALL_FACTORY_ADD |
-			       AGS_RECALL_FACTORY_PLAY |
-			       AGS_RECALL_FACTORY_RECALL),
+			       AGS_RECALL_FACTORY_PLAY),
 			      0);
   
     g_object_get(audio,
@@ -1302,18 +1305,18 @@ ags_lv2_bridge_map_recall(AgsMachine *machine)
 			      0, 0,
 			      (AGS_RECALL_FACTORY_INPUT |
 			       AGS_RECALL_FACTORY_ADD |
-			       AGS_RECALL_FACTORY_RECALL),
+			       AGS_RECALL_FACTORY_PLAY),
 			      0);
 
     g_object_get(audio,
-		 "recall", &start_recall,
+		 "play", &start_play,
 		 NULL);
 
-    recall = ags_recall_find_type(start_recall,
-				  AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
+    play = ags_recall_find_type(start_play,
+				AGS_TYPE_RECORD_MIDI_AUDIO_RUN);
 
-    if(recall != NULL){
-      recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(recall->data);
+    if(play != NULL){
+      recall_record_midi_audio_run = AGS_RECORD_MIDI_AUDIO_RUN(play->data);
     
       /* set dependency */
       g_object_set(G_OBJECT(recall_record_midi_audio_run),
@@ -1326,77 +1329,77 @@ ags_lv2_bridge_map_recall(AgsMachine *machine)
 		   NULL);
     }  
 
-    g_list_free(start_recall);
+    g_list_free(start_play);
 
-    /* ags-route-lv2 */
+    /* ags-play-notation */
     ags_recall_factory_create(audio,
 			      NULL, NULL,
-			      "ags-route-lv2",
+			      "ags-play-notation",
 			      0, 0,
 			      0, 0,
 			      (AGS_RECALL_FACTORY_INPUT |
 			       AGS_RECALL_FACTORY_ADD |
-			       AGS_RECALL_FACTORY_RECALL),
+			       AGS_RECALL_FACTORY_PLAY),
 			      0);
     
     g_object_get(audio,
-		 "recall", &start_recall,
+		 "play", &start_play,
 		 NULL);
 
-    recall = ags_recall_find_type(start_recall,
-				  AGS_TYPE_ROUTE_LV2_AUDIO_RUN);
+    play = ags_recall_find_type(start_play,
+				AGS_TYPE_PLAY_NOTATION_AUDIO_RUN);
 
-    if(recall != NULL){
-      recall_route_lv2_audio_run = AGS_ROUTE_LV2_AUDIO_RUN(recall->data);
+    if(play != NULL){
+      recall_notation_audio_run = AGS_PLAY_NOTATION_AUDIO_RUN(play->data);
 
       /* set dependency */
-      g_object_set(G_OBJECT(recall_route_lv2_audio_run),
+      g_object_set(G_OBJECT(recall_notation_audio_run),
 		   "delay-audio-run", play_delay_audio_run,
 		   NULL);
 
       /* set dependency */
-      g_object_set(G_OBJECT(recall_route_lv2_audio_run),
+      g_object_set(G_OBJECT(recall_notation_audio_run),
 		   "count-beats-audio-run", play_count_beats_audio_run,
 		   NULL);
-    }else{
-      recall_route_lv2_audio_run = NULL;
     }
+    
+    g_list_free(start_play);
 
-    g_list_free(start_recall);
-
-    /* ags-play-notation */
+    /* ags-route-lv2 */
     ags_recall_factory_create(audio,
 			      NULL, NULL,
-			      "ags-play-notation",
+			      "ags-route-lv2",
 			      0, 0,
 			      0, 0,
 			      (AGS_RECALL_FACTORY_INPUT |
 			       AGS_RECALL_FACTORY_ADD |
-			       AGS_RECALL_FACTORY_RECALL),
+			       AGS_RECALL_FACTORY_PLAY),
 			      0);
     
     g_object_get(audio,
-		 "recall", &start_recall,
+		 "play", &start_play,
 		 NULL);
 
-    recall = ags_recall_find_type(start_recall,
-				  AGS_TYPE_PLAY_NOTATION_AUDIO_RUN);
+    play = ags_recall_find_type(start_play,
+				AGS_TYPE_ROUTE_LV2_AUDIO_RUN);
 
-    if(recall != NULL){
-      recall_notation_audio_run = AGS_PLAY_NOTATION_AUDIO_RUN(recall->data);
+    if(play != NULL){
+      recall_route_lv2_audio_run = AGS_ROUTE_LV2_AUDIO_RUN(play->data);
 
       /* set dependency */
-      g_object_set(G_OBJECT(recall_notation_audio_run),
+      g_object_set(G_OBJECT(recall_route_lv2_audio_run),
 		   "delay-audio-run", play_delay_audio_run,
 		   NULL);
 
       /* set dependency */
-      g_object_set(G_OBJECT(recall_notation_audio_run),
+      g_object_set(G_OBJECT(recall_route_lv2_audio_run),
 		   "count-beats-audio-run", play_count_beats_audio_run,
 		   NULL);
+    }else{
+      recall_route_lv2_audio_run = NULL;
     }
-    
-    g_list_free(start_recall);
+
+    g_list_free(start_play);
   }
 
   /* depending on destination */
diff --git a/ags/X/machine/ags_matrix.c b/ags/X/machine/ags_matrix.c
index cb5c7ebe9..7b364d246 100644
--- a/ags/X/machine/ags_matrix.c
+++ b/ags/X/machine/ags_matrix.c
@@ -192,7 +192,8 @@ ags_matrix_init(AgsMatrix *matrix)
 			      AGS_AUDIO_ASYNC |
 			      AGS_AUDIO_OUTPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_RECYCLING));
-  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_SEQUENCER |
+  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_PLAYBACK |
+				      AGS_SOUND_ABILITY_SEQUENCER |
 				      AGS_SOUND_ABILITY_NOTATION));
   ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_PATTERN_MODE |
 					AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
@@ -488,12 +489,15 @@ ags_matrix_resize_pads(AgsMachine *machine, GType type,
       source = ags_channel_nth(source,
 			       pads_old);
 
-      if(source != NULL){
+      while(source != NULL){
 	AgsRecycling *recycling;
 	AgsAudioSignal *audio_signal;
 
 	GObject *output_soundcard;
 
+	ags_channel_set_ability_flags(source, (AGS_SOUND_ABILITY_SEQUENCER |
+						AGS_SOUND_ABILITY_NOTATION));
+	
 	g_object_get(audio,
 		     "output-soundcard", &output_soundcard,
 		     NULL);
@@ -510,6 +514,11 @@ ags_matrix_resize_pads(AgsMachine *machine, GType type,
 	audio_signal->flags |= AGS_AUDIO_SIGNAL_TEMPLATE;
 	ags_recycling_add_audio_signal(recycling,
 				       audio_signal);
+
+	/* iterate */
+	g_object_get(source,
+		     "next", &source,
+		     NULL);
       }
 
       if((AGS_MACHINE_MAPPED_RECALL & (machine->flags)) != 0){
diff --git a/ags/X/machine/ags_syncsynth.c b/ags/X/machine/ags_syncsynth.c
index 0077aa597..91c406f1f 100644
--- a/ags/X/machine/ags_syncsynth.c
+++ b/ags/X/machine/ags_syncsynth.c
@@ -194,7 +194,8 @@ ags_syncsynth_init(AgsSyncsynth *syncsynth)
 			      AGS_AUDIO_OUTPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_RECYCLING |
 			      AGS_AUDIO_INPUT_HAS_SYNTH));
-  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_NOTATION));
+  ags_audio_set_ability_flags(audio, (AGS_SOUND_ABILITY_PLAYBACK |
+				      AGS_SOUND_ABILITY_NOTATION));
   ags_audio_set_behaviour_flags(audio, (AGS_SOUND_BEHAVIOUR_REVERSE_MAPPING |
 					AGS_SOUND_BEHAVIOUR_DEFAULTS_TO_INPUT));
   g_object_set(audio,
@@ -779,6 +780,8 @@ ags_syncsynth_resize_audio_channels(AgsMachine *machine,
   AgsSyncsynth *syncsynth;
 
   AgsAudio *audio;
+  AgsChannel *output;
+  AgsChannel *channel, *next_pad;
 
   guint output_pads, input_pads;
 
@@ -790,9 +793,54 @@ ags_syncsynth_resize_audio_channels(AgsMachine *machine,
   g_object_get(audio,
 	       "output-pads", &output_pads,
 	       "input-pads", &input_pads,
+	       "output", &output,
 	       NULL);
 
   if(audio_channels > audio_channels_old){
+    /* AgsOutput */
+    channel = output;
+
+    while(channel != NULL){
+      /* get some fields */
+      g_object_get(channel,
+		   "next-pad", &next_pad,
+		   NULL);
+
+      channel = ags_channel_pad_nth(channel,
+				    audio_channels_old);
+
+      while(channel != next_pad){
+	AgsRecycling *recycling;
+	AgsAudioSignal *audio_signal;
+
+	GObject *output_soundcard;
+
+	ags_channel_set_ability_flags(channel, (AGS_SOUND_ABILITY_NOTATION));
+		
+	g_object_get(audio,
+		     "output-soundcard", &output_soundcard,
+		     NULL);
+
+	/* get recycling */
+	g_object_get(channel,
+		     "first-recycling", &recycling,
+		     NULL);
+
+	/* instantiate template audio signal */
+	audio_signal = ags_audio_signal_new(output_soundcard,
+					    (GObject *) recycling,
+					    NULL);
+	audio_signal->flags |= AGS_AUDIO_SIGNAL_TEMPLATE;
+	ags_recycling_add_audio_signal(recycling,
+				       audio_signal);
+
+	/* iterate */
+	g_object_get(channel,
+		     "next", &channel,
+		     NULL);
+      }
+    }
+
     if(ags_recall_global_get_rt_safe() ||
        ags_recall_global_get_performance_mode()){
       /* ags-copy */
@@ -933,15 +981,17 @@ ags_syncsynth_resize_pads(AgsMachine *machine, GType type,
 		   "output", &source,
 		   NULL);
 
-      source = ags_channel_nth(source,
-			       pads_old);
+      source = ags_channel_pad_nth(source,
+				   pads_old);
 
-      if(source != NULL){
+      while(source != NULL){
 	AgsRecycling *recycling;
 	AgsAudioSignal *audio_signal;
 
 	GObject *output_soundcard;
 	
+	ags_channel_set_ability_flags(source, (AGS_SOUND_ABILITY_NOTATION));
+	
 	g_object_get(audio,
 		     "output-soundcard", &output_soundcard,
 		     NULL);
@@ -958,6 +1008,11 @@ ags_syncsynth_resize_pads(AgsMachine *machine, GType type,
 	audio_signal->flags |= AGS_AUDIO_SIGNAL_TEMPLATE;
 	ags_recycling_add_audio_signal(recycling,
 				       audio_signal);
+
+	/* iterate */
+	g_object_get(source,
+		     "next-pad", &source,
+		     NULL);
       }
 
       if((AGS_MACHINE_MAPPED_RECALL & (machine->flags)) != 0){
@@ -1090,7 +1145,6 @@ ags_syncsynth_output_map_recall(AgsSyncsynth *syncsynth,
 {
   AgsAudio *audio;
 
-  guint input_pad_start;
   guint output_pads, input_pads;
   guint audio_channels;
 
@@ -1115,7 +1169,7 @@ ags_syncsynth_output_map_recall(AgsSyncsynth *syncsynth,
 			      NULL, NULL,
 			      "ags-copy",
 			      0, audio_channels, 
-			      input_pad_start, input_pads,
+			      0, input_pads,
 			      (AGS_RECALL_FACTORY_INPUT |
 			       AGS_RECALL_FACTORY_RECALL |
 			       AGS_RECALL_FACTORY_ADD),
@@ -1126,7 +1180,7 @@ ags_syncsynth_output_map_recall(AgsSyncsynth *syncsynth,
 			      NULL, NULL,
 			      "ags-buffer",
 			      0, audio_channels, 
-			      input_pad_start, input_pads,
+			      0, input_pads,
 			      (AGS_RECALL_FACTORY_INPUT |
 			       AGS_RECALL_FACTORY_RECALL |
 			       AGS_RECALL_FACTORY_ADD),
diff --git a/ags/audio/ags_audio.c b/ags/audio/ags_audio.c
index cf1ab6fc8..03a3b087b 100644
--- a/ags/audio/ags_audio.c
+++ b/ags/audio/ags_audio.c
@@ -9497,9 +9497,28 @@ ags_audio_real_play_recall(AgsAudio *audio,
       list_start = g_list_reverse(list_start);
   }
 
-  /* play  */
+  /* automate and play  */
   staging_flags = staging_mask & staging_flags;
   
+  if((AGS_SOUND_STAGING_AUTOMATE & (staging_flags)) != 0){
+    while(list != NULL){
+      recall = AGS_RECALL(list->data);
+      
+      /* play stages */
+      if(AGS_IS_RECALL_AUDIO(recall)){
+	ags_recall_set_staging_flags(recall,
+				     AGS_SOUND_STAGING_AUTOMATE);
+	ags_recall_unset_staging_flags(recall,
+				       AGS_SOUND_STAGING_AUTOMATE);
+      }
+      
+      list = list->next;
+    }
+  }
+
+  staging_flags &= (~AGS_SOUND_STAGING_AUTOMATE);
+  list = list_start;
+
   while((list = ags_recall_find_recycling_context(list,
 						  (GObject *) recycling_context)) != NULL){
     recall = AGS_RECALL(list->data);
diff --git a/ags/audio/ags_audio_signal.c b/ags/audio/ags_audio_signal.c
index bf5ce2397..0f780a38c 100644
--- a/ags/audio/ags_audio_signal.c
+++ b/ags/audio/ags_audio_signal.c
@@ -3034,9 +3034,9 @@ ags_audio_signal_feed(AgsAudioSignal *audio_signal,
 void
 ags_audio_signal_real_add_note(AgsAudioSignal *audio_signal,
 			       GObject *note)
-{  
+{
   if(g_list_find(audio_signal->note, note) == NULL){
-    g_object_ref(note);  
+    g_object_ref(note);
     audio_signal->note = g_list_prepend(audio_signal->note,
 					note);
   }
